toolchain/check/handle_class.cpp

```diff
@@ -526,6 +526,21 @@ auto HandleParseNode(Context& context, Parse::BaseDeclId node_id) -> bool {
     return true;
   }
 
+  // TODO: base must appear before virtual functions too
+  for (auto inst_id : context.inst_block_stack().PeekCurrentBlockContents()) {
```

Comment:
I worry that re-traversing the current block could be excessively costly. Do you think it might be sufficient to rephrase `ImplWithoutBase` to say there hasn't been a base declaration _yet_, and not have a separate diagnostic when we actually reach the base declaration?

Reply:
I suppose - it's sort of awkward that that's different from the field-before-base error (though that's meaningfully different than the impl-before-base, since impl-before-base can be identified as invalid at the impl, without having to wait for the base)

But if we want to error on virtual/abstract-before-base (for consistency? for detecting whether the virtual functions shadows something in the base (& should be impl instead), etc?), we can't do it at the virtual function and would need to do this second pass, or some other mechanism?

This will only traverse the members that appear before the base decl, right? Which `should` be relatively few? (though not guaranteed - since we're currently saying you can have any number of non-virtual functions declared before the base decl)

Reply:
> I suppose - it's sort of awkward that that's different from the field-before-base error (though that's meaningfully different than the impl-before-base, since impl-before-base can be identified as invalid at the impl, without having to wait for the base)

I see it as consistent: in both cases, we're diagnosing the error at the earliest point where it can be detected (and only there). The asymmetry between the two just reflects the asymmetry in the language rules: "`impl` must be preceded by `base`" is structurally different from "`base` must not be preceded by a field".

> But if we want to error on virtual/abstract-before-base (for consistency? for detecting whether the virtual functions shadows something in the base (& should be impl instead), etc?), we can't do it at the virtual function and would need to do this second pass, or some other mechanism?

I'm not sure of the best way to implement that. Possibly there's somewhere ephemeral where we could track whether we've seen any virtual/abstract methods yet?

> This will only traverse the members that appear before the base decl, right? Which `should` be relatively few? (though not guaranteed - since we're currently saying you can have any number of non-virtual functions declared before the base decl)

Yeah, I assume the prevailing style will be to put the `base` decl as close to the top of the class as possible, so the cost might not be that bad in practice. But even if performance isn't an issue, there's still the problem that IIUC this diagnostic can only catch issues that will also be caught (and caught earlier) by `ImplWithoutBase`. If that's the case, this diagnostic seems like unnecessary clutter.

Reply:
> > But if we want to error on virtual/abstract-before-base (for consistency? for detecting whether the virtual functions shadows something in the base (& should be impl instead), etc?), we can't do it at the virtual function and would need to do this second pass, or some other mechanism?
> 
> I'm not sure of the best way to implement that. Possibly there's somewhere ephemeral where we could track whether we've seen any virtual/abstract methods yet?

Yeah, this might be the best place to look further into - to answer some of the broader questions of vtable layout, etc, maybe we'll want some bigger change/more invasive representational features for dynamic classes. I've started a discussion ( https://github.com/carbon-language/carbon-lang/discussions/4671 ) to examine that.

Might hold off on this review (I can mark it abandoned, perhaps?) for now, then.